### PR TITLE
Set `MLFlowLogger` status to FAILED when training raises an error

### LINKED ISSUE #12291: `MLFlowLogger`'s status is "RUNNING" even after training failed
## 🐛 Bug

If a trainer with `MLFlowLogger` raises an error, the user should be able to see the MLflow's screen to check the training has been failed.
MLflow's status remains "RUNNING" even after `trainer.fit` raises an error in the current implementation, so the user cannot know whether the training is still in progress or failed.

#### Current behavior when training finished with an error: 

スクリーンショット 2022-03-10 21 41 18

#### Expected behavior: 

スクリーンショット 2022-03-10 21 31 23

### To Reproduce

```py
class CustomModel(BoringModel):
    def training_step(self, batch, batch_idx):
        super().training_step(batch, batch_idx)
        raise BaseException
trainer = Trainer(logger=MLFlowLogger("test"))
try:
    trainer.fit(CustomModel())
finally:
    print(trainer.logger.experiment.get_run(trainer.logger.run_id).info.status) # This should be 'FAILED'
```

cc @borda

### PR DESCRIPTION
**Pull request by ritsuki1227**
## What does this PR do?

If a trainer with MLFlowLogger raises an error, the user should be able to see the MLflow's screen to check the training has been failed.
MLflow's status remains "RUNNING" even after trainer.fit raises an error in the current implementation, so the user cannot know whether the training is still in progress or failed.

Fixes #12291

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ritsuki1227** [comment]
The commit has succeeded to fix the bug, but failed to pass an existing unit test, which I couldn't find a solution to.
Can anyone help me with the draft PR?

**Comment by ritsuki1227** [comment]
I would appreciate it if you approve running the workflows 🙂

File: pytorch_lightning/loggers/mlflow.py
Comment (awaelchli) [inline pytorch_lightning/loggers/mlflow.py:235]: an else is missing, which can lead to an undefined value for status. 
let's choose a default value, or not call set_terminated if status is neither finished nor failed.

File: pytorch_lightning/loggers/tensorboard.py
Comment (awaelchli) [inline pytorch_lightning/loggers/tensorboard.py:268]: I think we can still flush and close even if status is failed. Or is this not correct?

File: pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:734]: Great. I think we should follow up on this PR with a small refactor that defines an enum for statuses.

File: tests/loggers/test_mlflow.py
Comment (awaelchli) [inline tests/loggers/test_mlflow.py:245]: great test!

**Comment by awaelchli** [comment]
> I would appreciate it if you approve running the workflows 🙂

Most of team was OOO. Sorry about the delay. 

> The commit has succeeded to fix the bug, but failed to pass an existing unit test, which I couldn't find a solution to.
> Can anyone help me with the draft PR?

Which test was it?

**Comment by ritsuki1227** [comment]
@awaelchli 
Thank you for your review and modification!

> Which test was it?

It was `tests/utilities/test_cli.py::test_cli_distributed_save_config_callback`.
Although the latest implementation has passed the test, the failed test can be reproduced on commit dbbfef817a3fff52a4d0b98ef68ae39f03863f56.

File: pytorch_lightning/loggers/tensorboard.py
Comment (ritsuki1227) [reply pytorch_lightning/loggers/tensorboard.py:268]: `tests/utilities/test_cli.py::test_cli_distributed_save_config_callback` fails without the condition on commit dbbfef817a3fff52a4d0b98ef68ae39f03863f56.
I don't know why but somehow the problem was fixed on the current implementation.

**Comment by ritsuki1227** [comment]
@awaelchli @daniellepintz 
Do you have any update on this?
I understand some PRs may be pending because of the 1.6 release.

File: pytorch_lightning/loggers/tensorboard.py
Comment (carmocca) [reply pytorch_lightning/loggers/tensorboard.py:268]: Is this test still a problem? I am its author

**Comment by ritsuki1227** [comment]
@carmocca Is there anything I can help to merge the PR?

File: src/pytorch_lightning/loggers/tensorboard.py
Comment (carmocca) [inline src/pytorch_lightning/loggers/tensorboard.py:270]: Why don't we want to close it if it failed?

File: tests/tests_pytorch/loggers/test_mlflow.py
Comment (carmocca) [inline tests/tests_pytorch/loggers/test_mlflow.py:268]: You can raise the exception from an earlier hook like `on_fit_start`

**Review by carmocca — APPROVED** [review]
The changes look good to me. I'll merge master to see what CI thinks

File: src/pytorch_lightning/loggers/tensorboard.py
Comment (ritsuki1227) [reply src/pytorch_lightning/loggers/tensorboard.py]: @carmocca Hi, sorry for the late reply.
I've added the modification, but it makes `test_cli.py::test_cli_distributed_save_config_callback` fail.
I'd appreciate it if you help me with that.
FYI, I've found 10e133af6407463fe43951e9aa3153ff9b95c8c7 solves the problem as a workaround.

File: tests/tests_pytorch/loggers/test_mlflow.py
Comment (ritsuki1227) [reply tests/tests_pytorch/loggers/test_mlflow.py]: fixed the test to raise the exception on `on_fit_start`.

File: src/pytorch_lightning/loggers/tensorboard.py
Comment (awaelchli) [inline src/pytorch_lightning/loggers/tensorboard.py:109]: We shouldn't have any changes in tensorboard logger. Could you remove them?

File: src/pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:658]: If this gets moved here, then it shoud be removed from `Trainer._call_teardown_hook()`, right?

File: src/pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:666]: Note that with strategies like ddp_spawn, this won't do anything. On the main process, the logger has never done anything, so sending a "failed" signal will be wrong. 

There is currently no way we can handle this situation, and this is mostly the reason why only the "success" calls exist today. Sorry for the bad news :(

File: src/pytorch_lightning/loggers/tensorboard.py
Comment (carmocca) [reply src/pytorch_lightning/loggers/tensorboard.py:109]: See https://github.com/Lightning-AI/lightning/pull/12292#discussion_r959484188

File: src/pytorch_lightning/trainer/trainer.py
Comment (ritsuki1227) [reply src/pytorch_lightning/trainer/trainer.py:666]: Thank you for sharing. Is there anything I can do to help with the situation? I don't know If I have got the skills to do that.

Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:666]: A quick temporary fix is to do this in mlflowlogger.finalize:

```py
if self._mlflow_client is None:
    return

# else do your thing here to set the status to failed
```

This logic relies on the fact that _mlflow_client is only ever not None in the worker process. In the main process it is None unless the user called the logger manually. 

This "hack" is not ideal, because it shifts responsibility to the logger that should ideally be handled by probably the launcher (i.e., only calling the finalize on the right process). 
This mechanism to process exceptions before they bubble up to the main process would be necessary to have for handling this properly. It needs some brainstorming, so probably the next step is to open a GH issue. 

For now, let's try to go with my suggestion and try to land the PR, what do you think? 

Note the above snippet would have to be adapted for each logger that implements finalize, i.e., checking whether `self._experiment is None`.

Comment (ritsuki1227) [reply src/pytorch_lightning/trainer/trainer.py]: @awaelchli 
> checking whether self._experiment is None.

In addition to the fact that objects to be checked were different for each logger, it is unclear how other loggers behave when `finalize("failed")` is called.
So I fixed it so that `finalize("failed")` is called on an exception only when the logger is mlfow. Can you review the new commits?
A quick overview of the changes is as follows:

trainer.py:
```py
def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs: Any) -> Any:
        ...
        except BaseException as exception:
            ...
            for logger in self.loggers:
                logger.finalize_when_exception()
```

logger.py:
```py
class Logger(ABC):
    ...
    def finalize_when_exception(self) -> None:
        pass
```

mlflow.py: 
```py
class MLFlowLogger(Logger):
    ...
    def finalize_when_exception(self) -> None:
        super().finalize_when_exception()
        if self._mlflow_client is None:
            return
        self.finalize("failed")
```

Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py]: > In addition to the fact that objects to be checked were different for each logger, it is unclear how other loggers behave when finalize("failed") is called.

If they don't do anything now, then we leave it like that. 

I wouldn't introduce any new api here, especially not in the base class when it's just affecting one logger. Did my [suggestion](https://github.com/Lightning-AI/lightning/pull/12292#discussion_r961550279) not work?
Happy to help and get hands-on here. Should I push a commit? I can take care of the other loggers.

Comment (ritsuki1227) [reply src/pytorch_lightning/trainer/trainer.py]: @awaelchli It would be helpful if you push a commit or give me some advice!

> Did my https://github.com/Lightning-AI/lightning/pull/12292#discussion_r961550279 not work?

If you fix `tensorboard.py` like this: 

```py
    @rank_zero_only
    def finalize(self, status: str) -> None:
        if self._experiment is None:
            return
        self.experiment.flush()
        self.experiment.close()
        self.save()
```

Then `tests/tests_pytorch/loggers/test_tensorboard.py::test_tensorboard_save_hparams_to_yaml_once` fails: 

```py
    def test_tensorboard_save_hparams_to_yaml_once(tmpdir):
        model = BoringModel()
        logger = TensorBoardLogger(save_dir=tmpdir, default_hp_metric=False)
        trainer = Trainer(max_steps=1, default_root_dir=tmpdir, logger=logger)
        assert trainer.log_dir == trainer.logger.log_dir
        trainer.fit(model)
    
        hparams_file = "hparams.yaml"
>       assert os.path.isfile(os.path.join(trainer.log_dir, hparams_file))
E       AssertionError: assert False
E        +  where False = <function isfile at 0x1059b8670>('/private/var/folders/01/y61dt9vx52lds6ylr3975mbc0000gq/T/pytest-of-ritsuki.yamada/pytest-93/test_tensorboard_save_hparams_0/lightning_logs/version_0/hparams.yaml')
E        +    where <function isfile at 0x1059b8670> = <module 'posixpath' from '/Users/ritsuki.yamada/.pyenv/versions/3.9.13/lib/python3.9/posixpath.py'>.isfile
E        +      where <module 'posixpath' from '/Users/ritsuki.yamada/.pyenv/versions/3.9.13/lib/python3.9/posixpath.py'> = os.path
E        +    and   '/private/var/folders/01/y61dt9vx52lds6ylr3975mbc0000gq/T/pytest-of-ritsuki.yamada/pytest-93/test_tensorboard_save_hparams_0/lightning_logs/version_0/hparams.yaml' = <function join at 0x1059b8e50>('/private/var/folders/01/y61dt9vx52lds6ylr3975mbc0000gq/T/pytest-of-ritsuki.yamada/pytest-93/test_tensorboard_save_hparams_0/lightning_logs/version_0', 'hparams.yaml')
E        +      where <function join at 0x1059b8e50> = <module 'posixpath' from '/Users/ritsuki.yamada/.pyenv/versions/3.9.13/lib/python3.9/posixpath.py'>.join
E        +        where <module 'posixpath' from '/Users/ritsuki.yamada/.pyenv/versions/3.9.13/lib/python3.9/posixpath.py'> = os.path
E        +      and   '/private/var/folders/01/y61dt9vx52lds6ylr3975mbc0000gq/T/pytest-of-ritsuki.yamada/pytest-93/test_tensorboard_save_hparams_0/lightning_logs/version_0' = <pytorch_lightning.trainer.trainer.Trainer object at 0x130584dc0>.log_dir

tests/tests_pytorch/loggers/test_tensorboard.py:291: AssertionError
```

Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:656]: @ritsuki1227 I noticed this too. This was because the save() in finalize of TensorBoard was expected to be called, even when no self.experiment is initialized. This is because the saving of hparams happens independently of the tensorboard summary writer. Now, I implemented it so that it behaves the same as on master. 

Otherwise, I took care of:
- Other loggers
- Improving the tests

I let the CI run now and check back later. If other tests fail, I will take another look.

**Comment by awaelchli** [comment]
@ritsuki1227 I created a fix in #14762 to unblock this PR for the remaining issues.

File: src/pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:648]: Right now, KeyboardInterrupt is the same as "failed". Perhaps we want to introduce something like "stopped" in the future. How the loggers react here depends on how the third-party implementations expect things to be reported when an experiment gets "stopped" or "interrupted" in a controlled way.

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/loggers/wandb.py
Comment (rohitgr7) [inline src/pytorch_lightning/loggers/wandb.py:555]: experiment is initialized during init, so will it be None on the main process?

**Review by rohitgr7 — APPROVED** [review]
mind add a changelog?

File: src/pytorch_lightning/loggers/wandb.py
Comment (awaelchli) [reply src/pytorch_lightning/loggers/wandb.py:555]: You're right! On wandb >= 0.12.10 that's the case. 
I changed the logic now to keep the behavior on master, which is: only upload checkpoints on success

**Comment by awaelchli** [comment]
@ritsuki1227 before we merge, would you mind checking again your MLFlow use case, that the status is reported as you expect? Thank you!

**Comment by ritsuki1227** [comment]
@awaelchli Thank you for your commits! I've checked mlflow logger worked correctly on the use case.

**Comment by awaelchli** [comment]
Awesome, let's get it in then! Thanks a lot for your help in making the integration with PL better!

**Pull request merged**
